### ISSUE TITLE
[BUG] Questions about 'get_mask_subset_prob' and MLM in soundstorm.py

### ISSUE THREAD

**Issue by dbkest** [lucidrains/soundstorm-pytorch#29]
hi, great project. when reading your training code in soundstorm.py（commit 264a1f2）, there are 2 places are hard for me to understand, could you please explain for me, thanks. 

**que1**: get_mask_subset_prob in line86, I understand that this function has two purposes: 1. Randomly sample from elements that are True in the mask to create a subset_mask. 2. The elements that are True in the subset_mask should also be True in the corresponding positions in the mask. However, it seems that it cannot achieve the above purpose 1. Please refer to the following example:
index:       0      1    2   3   4     5     6   7 
mask:        1     1     1    1    1     1     0   0.  prob=0.3 min_mask=0 -> num_to_mask = 1.8 num_padding=2
logits:       0.2 0.4 0.1 0.3 0.6 0.5 -1 -1  #line98
randperm:  6   7    2.   0.  3.    1     5  4  #line100
randperm:  5   6    1   -1   2.    0    4  3  #line103
sub_mask:  0   0    1     1   0     1    0   0.  #line 105 
sub_mask:  0   0    1     1   0     1    0   0.  #line 106

so it cannot achieve the above purpose 1: sum(sub_mask) > num_to_mask; 
Another problem is  that the elements in the range of 0~ num_padding in the sub_mask are most likely to be False, when num_padding > 0, because the larger padding elements' indices will always be sorted to the front in #line 100, which result in the model only learning the token-embedding from the latter half of the training data.

I attempted to fix the above issue, the code is as follows：

def get_mask_subset_prob(
    mask: Tensor,
    prob: Union[float, Tensor],
    min_mask: int = 0
):
    batch, seq, device = *mask.shape, mask.device

    if isinstance(prob, Tensor):
        prob = rearrange(prob, 'b -> b 1')

    num_to_mask = (mask.sum(dim = -1, keepdim = True) * prob).clamp(min = min_mask)
    logits = torch.rand((batch, seq), device = device)
    # logits = logits.masked_fill(~mask, -1)
    logits = logits.masked_fill(~mask, 2) #mod: place padding elements to end position

    randperm = logits.argsort(dim = -1)
    randperm = randperm.argsort(dim = -1).float() #mod: randperm should be score instead of origin index

    #num_padding = (~mask).sum(dim = -1, keepdim = True) # mod: delete
    #randperm -= num_padding #mod: delete

    subset_mask = randperm < num_to_mask
    #subset_mask.masked_fill_(~mask, False) #mod: delete
    return subset_mask

**que2**: I do not quite understand the role of introducing MLM in lines 1129 to 1140. In lines 1129 to 1133, randomly selecting some elements not to be replaced seems to only reduce the number of masked tokens. The tokens randomly replaced in lines 1135 to 1140 are only used as input and are not predicted to calculate loss. It's like creating errors in the input and having the model predict the masked token based on erroneous input. I guess this is to minimize the gap between training and inference, because SoundStorm will perform multi-step inference at level 0, and some tokens predicted in historical steps may be wrong?

I sincerely await your reply.

**Comment by lucidrains** [comment]
@dbkest yes you are right for number one, there is an error

could you check that 0.4.2 lines up?

as for the MLM, it is just a random research idea I had for improving self conditioning. it is probably safe to just turn it off

**Comment by lbdave94** [comment]
@dbkest sorry, what do you mean with "_it seems that it cannot achieve the above purpose 1_: sum(sub_mask) > num_to_mask;" ? In the example that you posted the 2 conditions are satisfied:

1. _The elements that are True in the subset_mask should also be True in the corresponding positions in the mask_
2. _sum(sub_mask) > num_to_mask_ (you have 3 True, where num_to_mask is 1.8)

While I'm agreed about the problem to moving the padding at the start.

**Comment by lbdave94** [comment]
@dbkest sorry, what do you mean with "_it cannot achieve the above purpose 1: sum(sub_mask) > num_to_mask_" ? In the example you posted:

1. _The elements that are True in the subset_mask should also be True in the corresponding positions in the mask_
2. _sum(sub_mask) > num_to_mask_ (You have 3 True, where num_to_mask is 1.8)

While I'm agree on the problem to move padding at the start.

**Comment by lucidrains** [comment]
@lbdave94 it should be fixed.. i had some wrong mental model about `.argsort` unfortunately and this bug appeared in two or three repos 🙏

**Comment by lucidrains** [comment]
@lbdave94 the issue isn't that it doesn't produce the correct number of subset mask `True` elements, but that they were all sorted to one end of the sequence or the other

**Comment by dbkest** [comment]
> @lbdave94 the issue isn't that it doesn't produce the correct number of subset mask `True` elements, but that they were all sorted to one end of the sequence or the other
@lucidrains I think you are right. Thanks for your reply.

**Comment by lucidrains** [comment]
@dbkest well, after this project, at least i will never forget about the double `.argsort` trick.
